pennylane/transforms/qmc.py

```diff
@@ -0,0 +1,353 @@
+        scales as :math:`2^{n}`. However, due to the use of quantum phase estimation, the error
+        :math:`\epsilon` scales as :math:`\mathcal{O}(2^{-n})`. Hence,
+
+        .. math::
+
+            N = \mathcal{O}\left(\frac{1}{\epsilon}\right).
+
+        This scaling can be compared to standard Monte Carlo estimation, where :math:`N` samples are
+        generated from the probability distribution and the average over :math:`f` is taken. In that
+        case,
+
+        .. math::
+
+            N =  \mathcal{O}\left(\frac{1}{\epsilon^{2}}\right).
+
+        Hence, the quantum Monte Carlo algorithm has a quadratically improved time complexity with
+        :math:`N`.
+
+        **Example**
+
+        Consider a standard normal distribution :math:`p(x)` and a function
+        :math:`f(x) = \sin ^{2} (x)`. The expectation value of :math:`f(x)` is
+        :math:`\int_{-\infty}^{\infty}f(x)p(x) \approx 0.432332`. This number can be approximated by
+        discretizing the problem and using the quantum Monte Carlo algorithm.
+
+        First, the problem is discretized:
+
+        .. code-block:: python
+
+            from scipy.stats import norm
+
+            m = 5
+            M = 2 ** m
+
+            xmax = np.pi  # bound to region [-pi, pi]
+            xs = np.linspace(-xmax, xmax, M)
+
+            probs = np.array([norm().pdf(x) for x in xs])
+            probs /= np.sum(probs)
+
+            func = lambda i: np.sin(xs[i]) ** 2
+            r_rotations = np.array([2 * np.arcsin(np.sqrt(func(i))) for i in range(M)])
+
+        The ``quantum_monte_carlo`` transform can then be used:
+
+        .. code-block::
+
+            from pennylane.templates.state_preparations.mottonen import (
+                _uniform_rotation_dagger as r_unitary,
+            )
+
+            n = 6
+            N = 2 ** n
+
+            a_wires = range(m)
+            wires = range(m + 1)
+            target_wire = m
+            estimation_wires = range(m + 1, n + m + 1)
+
+            dev = qml.device("default.qubit", wires=(n + m + 1))
+
+            def fn():
+                qml.templates.MottonenStatePreparation(np.sqrt(probs), wires=a_wires)
+                r_unitary(qml.RY, r_rotations, control_wires=a_wires[::-1], target_wire=target_wire)
+
+            @qml.qnode(dev)
+            def qmc():
+                qml.quantum_monte_carlo(fn, wires, target_wire, estimation_wires)()
+                return qml.probs(estimation_wires)
+
+            phase_estimated = np.argmax(qmc()[:int(N / 2)]) / N
+
+        The estimated value can be retrieved using the formula :math:`\mu = (1-\cos(\pi \theta))/2`
+
+        >>> (1 - np.cos(np.pi * phase_estimated)) / 2
+        0.42663476277231915
+
+        It is also possible to explore the resources required to perform the quantum Monte Carlo
+        algorithm
+
+        >>> qtape = qmc.qtape.expand(depth=1)
+        >>> qtape.get_resources()
```

Comment:
Whaaat!?? :exploding_head: :exploding_head: :exploding_head: :exploding_head: 
I had no idea we could do this. So cool!

Reply:
Yes!! But don't advertise this yet as we don't want this to be user facing. Instead, we will be going with the following approaches:

https://github.com/PennyLaneAI/pennylane/issues/1293
https://github.com/PennyLaneAI/adrs/pull/13

In particular, the ability to do 

```pycon
>>> qml.specs(qmc)(*params) # get all specs
>>> qml.depth(qml)(*params) # get a specific spec
```

Reply:
Cool! @josh146 do you recommend removing this part then? On the other hand, it's a shame to hide the gate count.

Reply:
Good question. I would say - leave it for now, but we need to remember to update it when the preferred UI changes 🙂